### PR TITLE
[MAINTENANCE] Make DataAssistantResult.usage_statistics_handler private (in order to optimize auto-complete for ease of use)

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* By changing the `DataAssistantResult.usage_statistics_handler` property name to `DataAssistantResult._usage_statistics_handler` the `DataAssistantResult` will be less cluttered in `Jupyter` notebooks when using the "autocomplete" feature.
* This requires a slightly more advanced usage of the Python dataclass library.
* Ensure that existing tests pass.
* Ensure that autocomplete in a `Jupyter` notebook does not display either `usage_statistics_handler` or `_usage_statistics_handler` when hitting "Tab" on `result.` (where `result: DataAssistantResult`).

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- JIRA: GREAT-761/GREAT-1101
-
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

**Review by talagluck — APPROVED** [review]
LGTM!

File: great_expectations/rule_based_profiler/data_assistant/data_assistant.py [hunk 1]
```diff
@@ -463,7 +463,7 @@ def run(
         data_assistant_result: DataAssistantResult = DataAssistantResult(
             _batch_id_to_batch_identifier_display_name_map=self._batch_id_to_batch_identifier_display_name_map(),
             execution_time=0.0,
-            usage_statistics_handler=usage_statistics_handler,
+            _usage_statistics_handler=usage_statistics_handler,
```
Comment (donaldheppner) [inline great_expectations/rule_based_profiler/data_assistant/data_assistant.py:466]: we should not change the name of the parameter in the method

File: great_expectations/rule_based_profiler/rule_based_profiler_result.py [hunk 2]
```diff
@@ -70,7 +71,7 @@ def to_dict(self) -> dict:
             ],
             "citation": self.citation,
             "execution_time": self.execution_time,
-            "usage_statistics_handler": self.usage_statistics_handler.__class__.__name__,
+            "_usage_statistics_handler": self._usage_statistics_handler.__class__.__name__,
```
Comment (donaldheppner) [inline great_expectations/rule_based_profiler/rule_based_profiler_result.py:74]: Why were we ever including this in to_dict? there should be no reason to ever show this; please remove

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py [hunk 3]
```diff
@@ -138,7 +138,7 @@ class DataAssistantResult(SerializableDictDot):
         "expectation_configurations",
         "citation",
         "execution_time",
-        "usage_statistics_handler",
+        "_usage_statistics_handler",
```
Comment (donaldheppner) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:141]: this field should not be serialized, and need not be included in this list

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py [hunk 4]
```diff
@@ -161,7 +161,8 @@ class DataAssistantResult(SerializableDictDot):
     execution_time: Optional[
         float
     ] = None  # Overall DataAssistant execution time (in seconds).
-    usage_statistics_handler: Optional[UsageStatisticsHandler] = None
+    # Reference to "UsageStatisticsHandler" object for this "DataAssistantResult" object (if configured).
+    _usage_statistics_handler: Optional[UsageStatisticsHandler] = field(default=None)
```
Comment (donaldheppner) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:165]: not required for serialization

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py [hunk 5]
```diff
@@ -197,7 +198,7 @@ def to_dict(self) -> dict:
             ],
             "citation": convert_to_json_serializable(data=self.citation),
             "execution_time": convert_to_json_serializable(data=self.execution_time),
-            "usage_statistics_handler": self.usage_statistics_handler.__class__.__name__,
+            "_usage_statistics_handler": self._usage_statistics_handler.__class__.__name__,
```
Comment (donaldheppner) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:201]: not needed in to_dict

**Review by donaldheppner — COMMENTED** [review]
This code considers usage_statistics_handler as a config element the same as variables and rules. This is not the case. It should never be serialized and should not be part of schema; it is only present at runtime

File: great_expectations/rule_based_profiler/rule_based_profiler_result.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/rule_based_profiler_result.py:74]: @donaldheppner The reason is "informational" for printing and/or troubleshooting purposes.  I am happy to make the names without the underscores.  We can of course remove "usage_statistics_handler" (with or without the underscore) from `to_dict()` entirely, but then we would lose knowing that the object actually had the usage stats handler, which, seems a small disadvantage compared to knowing its name.  Happy to discuss (easy change).  Thanks.

**Comment by alexsherstinsky** [comment]
@donaldheppner This is addressed in #5571 -- thanks!

**Pull request merged**